File: Cpp/0999_available_captures_for_rook.cpp
// https://leetcode.cn/problems/available-captures-for-rook/

class Solution {
public:
    int numRookCaptures(vector<vector<char>>& board) {
        int res = 0;
        vector<pair<int, int>> dirs = {{0,1}, {0,-1}, {1,0}, {-1,0}};
        for (int i = 0; i < 8; ++i) {
            for (int j = 0; j < 8; ++j) {
                if (board[i][j] != 'R') continue;
                for (auto d: dirs) {
                    int i0 = i + d.first, j0 = j + d.second;
                    while (i0 >= 0 && i0 < 8 && j0 >= 0 && j0 < 8) {
                        if (board[i0][j0] == 'p') {
                            res += 1;
                            break;
                        }
                        else if (board[i0][j0] == 'B') {
                            break;
                        }
                        i0 += d.first;
                        j0 += d.second;
                    }
                }
            }
        }
        return res;
    }
};

File: Python3/0999_available_captures_for_rook.py
# https://leetcode.cn/problems/available-captures-for-rook/

class Solution:
    def numRookCaptures(self, board):
        """
        Inputs:
            board: list[list[str]]
        Outputs:
            res: int
        """        
        res = 0
        dirs = [[-1, 0], [1, 0], [0, -1], [0, 1]]
        for i in range(8):
            for j in range(8):
                if board[i][j] != "R":
                    continue
                for d in dirs:
                    i0, j0 = i+d[0], j+d[1]
                    while i0 >= 0 and i0 < 8 and j0 >= 0 and j0 < 8:
                        if board[i0][j0] == "p":
                            res += 1
                            break
                        elif board[i0][j0] == "B":
                            break
                        
                        i0 += d[0]
                        j0 += d[1]
        return res